presto-main/src/main/java/com/facebook/presto/operator/ExplainAnalyzeOperator.java

```diff
@@ -134,13 +136,26 @@ public Page getOutput()
         if (!finishing) {
             return null;
         }
-        outputConsumed = true;
+
         QueryInfo queryInfo = queryPerformanceFetcher.getQueryInfo(operatorContext.getDriverContext().getTaskId().getQueryId());
+
         // Skip the output stage, since that's the one that has the ExplainAnalyzeOperator itself
         checkState(queryInfo.getOutputStage().isPresent(), "Output stage is missing");
+
+        // Check if all stages, except the output stage, have the final StageInfo
+        boolean finalStageInfo = getAllStages(queryInfo.getOutputStage()).stream()
```

Comment:
In this case, the operator is able to process (and discard) input until the source of this operator (an ExchangeOperator) is finished.  This operator blocks and unblocks as the exchange blocks and unblocks, but once the exchange is finished it no longer blocks and this operator is the only thing holding up the completions of the query.  Without a future to wait on, I think the best we can do is a loop with a sleep.  Something like:
```
if (!hasFinalInfo) {
   sleep(100ms);
}
if (!hasFinalInfo) {
   return null;
}
return null
```
This will slow down the loop at the cost of a single thread, but that isn't a huge deal for a cluster of machines.